src/borrow_tracker/tree_borrows/perms.rs

```diff
@@ -267,6 +267,19 @@ impl Permission {
         transition::perform_access(kind, rel_pos, old_state, protected)
             .map(|new_state| PermTransition { from: old_state, to: new_state })
     }
+
+    /// During a provenance GC, we want to compact the tree.
+    /// For this, we want to merge nodes upwards if they have a singleton parent.
+    /// But we need to be careful: If the parent is Frozen, and the child is Reserved,
+    /// we can not do such a merge. In general, such a merge is possible if the parent
+    /// is not protected, and if the child is reachable from the parent. This
+    /// corresponds to the `partial_cmp` order on permissions.
+    pub fn can_be_replaced_by_child(self, child: Self) -> bool {
+        match self.inner.partial_cmp(&child.inner) {
+            Some(Ordering::Less) | Some(Ordering::Equal) => true,
+            Some(Ordering::Greater) | None => false,
+        }
+    }
```

Comment:
This is the magic function. I think that this is sound, i.e. if it returns `true`, then the child can indeed replace the parent.

Here is an example: The parent is `Active`, and the child is `Frozen`. On a foreign write, both go to `Disabled`. On a foreign read, both go to `Frozen`. On a child read, both remain, and the `Frozen` tag blocks child writes. So the `Active` has no effect.